tests/integration/high_availability/test_self_healing.py

```diff
@@ -315,3 +319,99 @@ async def test_network_cut_affecting_an_instance(
+
+
+@pytest.mark.order(2)
+@pytest.mark.abort_on_fail
+@pytest.mark.self_healing_tests
+async def test_graceful_full_cluster_crash_test(
+    ops_test: OpsTest, continuous_writes, restart_policy
+) -> None:
+    """Test to send SIGTERM to all units and then ensure that the cluster recovers."""
+    mysql_application_name, application_name = await high_availability_test_setup(ops_test)
+
+    time.sleep(30)
+
+    await ensure_all_units_continuous_writes_incrementing(ops_test)
+
+    assert await ensure_n_online_mysql_members(
+        ops_test, 3
+    ), "The deployed mysql application does not have three online nodes"
+
+    mysql_units = ops_test.model.applications[mysql_application_name].units
+
+    unit_mysqld_pids = {}
+    for unit in mysql_units:
+        pid = await get_process_pid(ops_test, unit.name, MYSQL_CONTAINER_NAME, MYSQLD_PROCESS_NAME)
+        assert pid > 1
+
+        unit_mysqld_pids[unit.name] = pid
+
+    written_value = await get_max_written_value_in_database(ops_test, mysql_units[0])
+
+    for unit in mysql_units:
+        await send_signal_to_pod_container_process(
+            ops_test, unit.name, MYSQL_CONTAINER_NAME, MYSQLD_PROCESS_NAME, "SIGTERM"
+        )
+
+    for attempt in Retrying(stop=stop_after_delay(60), wait=wait_fixed(10)):
+        with attempt:
+            for unit in mysql_units:
+                await ensure_process_not_running(
+                    ops_test, unit.name, MYSQL_CONTAINER_NAME, MYSQLD_PROCESS_NAME
+                )
+
+    async with ops_test.fast_forward():
+        await ops_test.model.block_until(
+            lambda: ops_test.model.applications[mysql_application_name].status != "active",
+            timeout=15 * 60,
+        )
+
+    # restart the cluster
+    await update_pebble_plan(
+        ops_test, {"on-failure": "restart", "on-success": "restart"}, mysql_application_name
+    )
+
+    async with ops_test.fast_forward():
+        # wait for model to stabilize, and all members to recover
+        await ops_test.model.wait_for_idle(
+            apps=[mysql_application_name],
+            status="active",
+            raise_on_blocked=False,
+            timeout=TIMEOUT,
+            idle_period=30,
+        )
+
+    for unit in mysql_units:
+        new_pid = await get_process_pid(
+            ops_test, unit.name, MYSQL_CONTAINER_NAME, MYSQLD_PROCESS_NAME
+        )
+        assert new_pid > 1
+        assert new_pid > unit_mysqld_pids[unit.name]
+
+    cluster_status = await get_cluster_status(ops_test, mysql_units[0])
+    assert all(
+        [
+            member["status"] == "online"
+            for member in cluster_status["defaultreplicaset"]["topology"].values()
+        ]
+    )
+
+    # The full cluster crash causes the continuous writes application to raise an error
+    # Thus we reset continuous writes and ensure that they are incrementing on all units
+    application_unit = ops_test.model.applications[application_name].units[0]
```

Comment:
I thinking about two approaches here (yet to be implemented), instead of this:
- update relation data `endpoints` faster
- fix continuous write to pickup the update faster

either way, your solution is good